text/trace/0000-error_flagging.md

```diff
@@ -0,0 +1,86 @@
+# Error Flagging with Status Codes
+This proposal adds two status codes explicitly for use as overrides by the end user, and proposes a canonical mapping of semantic conventions to status codes. This clarifies how error reporting should work in OpenTelemetry.
+ 
+## Motivation
+Error reporting is a fundamental use case for distributed tracing. While we prefer that error flagging occurs within analysis tools, and not within instrumentation plugins, a number of currently supported analysis tools and protocols rely on the existence of an explicit error flag reported from instrumentation. In OpenTelmetry, the error flag is called "status codes."
+ 
+However, there is confusion over the mapping of semantic conventions to status codes, and concern over the subjective nature of errors. Which network failures count as an error? Are 404s an error? The answer is dependent on the situation.
+ 
+There is one major exception. Both application developers and operators have a deep understanding of what constitutes an error in their system. OpenTelemetry must provide a way for these users to control error flagging, and explicitly indicate both when a span should and should not be counted as an error.
+A second exception is supporting analysis tools which require explicit error flagging in the data which they receive. In this case, an operator must be able to apply an error flagging schema at some point during the OTLP data processing pipeline.
+ 
+## Explanation
+The following changes add several missing features required for proper error reporting, and are completely backwards compatible with OpenTelemetry today.
+ 
+### Status Codes
+The following status codes are added to our current  schema.
+ 
+* `DEFAULT` No status has been set. Any errors must be detected by the analysis tool. (This replaces `OK`.)
+* `ERROR` Instrumentation has marked a span as an error. (This replaces `UNKNOWN`.)
+* `OK_OVERRIDE` The user has provided an override. The span should NOT be flagged as an error, regardless of other analysis.
+* `ERROR_OVERRIDE` The user has provided an override. The span SHOULD be flagged as an error, regardless of other analysis.
```

Comment:
If I understand correctly one of the differences between `ERROR` and `ERROR_OVERRIDE` is that `ERROR` is set by instrumentation and may be ignored by analysis tools, while `ERROR_OVERRIDE` is set by application and should be honored.

This seems to hint that we believe that the opinion of instrumentation developers is less reliable and may be ignored but the opinion of application developers is more important and should be honored.

Is this a correct reading of the proposal?

If so then the rationale is not completely clear. Why do we think (if we think so) the opinion of instrumentation developers is less reliable or less important than opinion of application developers?

I can see that we aim to resolve the conflict of 3 opinions here: 1) the instrumentation developer, 2) the application developer and 3) the analysis tool (and the last one may be a composite opinion influenced by the analysis tool developer and the analysis tool user if the tool is configurable).

I feel that the proposal would be stronger if it provided the rationale for choices made on how the conflict is resolved.

Reply:
Sorry, the intent is definitely not to disparage the instrumentation writers! The reason we need to differentiate between instrumentation and the end user is because most instrumentation is generic. Most instrumentation will only supply a standard mapping of semantics to status codes. On the other hand, statuses set by application developers and operators is much more service and deployment specific. Even if systems have their own schema, and ignore some or all statuses set by instrumentation, they will will still be interested in this information, as it is novel and very relevant. 

Does that sound reasonable?

Reply:
The way I see it, instrumentation developers make their best effort attempt at accurately reporting facts of what happened during instrumented calls.  The instrumentation developer must be far more strict about being subjective about what constitutes an error whereas the application developer has more freedom to be subjective.  An application developer, who is using an instrumentation has more context to make explicit decisions that may go against the generic (more context-less) indications an instrumentation is setting for error conditions.  In other words, what @tedsuo replied with is 100% reasonable and spot on to me.

Reply:
@tedsuo sounds good. If possible please add a couple words to the OTEP to make it clear.

Reply:
Thanks. I updated several sections.

> There is one other missing piece, required for proper error flagging. Both application developers and operators have a deep understanding of what constitutes an error in their system. OpenTelemetry must provide a way for these users to control error flagging, and explicitly indicate that it is the end user setting the status code, and not instrumentation plugins. In these specific cases, the error flagging is known to be correct: the end user has decided the status of the span, and they do not want another interpretation. 

> Note that analysis tools MAY disregard Span Status, and do their own error analysis. There is no requirement that the status code is respected, even when `user_override` is set. However, it is strongly suggested that analysis tools SHOULD pay attention to the status code when `user_override` is set, as it represents a subjective decision made by either the operator or application developer.

Reply:
Please let me know if it is clear enough now.